libs/cardano-ledger-core/src/Cardano/Ledger/TerseTools.hs

```diff
@@ -0,0 +1,107 @@
+  terse (a, b, c) = "(" ++ terse a ++ ", " ++ terse b ++ ", " ++ terse c ++ ")"
+
+caseKey :: Case p b -> p
+caseKey (OnLeft (k, _)) = k
+caseKey (OnRight (k, _)) = k
+caseKey (SameKey (k, _, _)) = k
+
+-- | we assume the lists are lexigraphically sorted
+differences :: (Ord a, Eq b) => [(a, b)] -> [(a, b)] -> [Case a b]
+differences [] [] = []
+differences xs [] = (map OnLeft xs)
+differences [] ys = (map OnRight ys)
+differences ((a1, b1) : xs) ((a2, b2) : ys) =
+  case compare a1 a2 of
+    EQ -> if b1 == b2 then differences xs ys else (SameKey (a1, b1, b2)) : differences xs ys
+    LT -> (OnLeft (a1, b1)) : differences xs ((a2, b2) : ys)
+    GT -> (OnRight (a2, b2)) : differences ((a1, b1) : xs) ys
+
+mapdiffs :: (Ord a, Eq b) => Map.Map a b -> Map.Map a b -> [Case a b]
+mapdiffs mp1 mp2 = differences (Map.toAscList mp1) (Map.toAscList mp2)
+
+terselist :: Terse a => [Char] -> [a] -> [Char]
+terselist message xs = "\n" ++ message ++ "\n" ++ unlines (map terse xs)
+
+terselistfilter :: Terse a => [Char] -> (a -> Bool) -> [a] -> [Char]
+terselistfilter message p xs = "\n" ++ message ++ "\n" ++ unlines (map terse (filter p xs))
+
+tersemap :: (Terse k, Terse a) => [Char] -> Map.Map k a -> [Char]
+tersemap message mp = terselist message (Map.toAscList mp)
+
+tersemapfilter :: (Terse k, Terse a) => [Char] -> (a -> Bool) -> Map.Map k a -> [Char]
+tersemapfilter message p mp = terselistfilter message (\(_, a) -> p a) (Map.toAscList mp)
+
+tersemapdiffs :: (Terse a, Terse b, Ord a, Eq b) => String -> Map.Map a b -> Map.Map a b -> [Char]
+tersemapdiffs message mp1 mp2 = terselist message (mapdiffs mp1 mp2)
+
+instance Terse (Addr crypto) where
+  terse (Addr _net cred1 (StakeRefBase cred2)) = "Addr (" ++ terse cred1 ++ ") (" ++ terse cred2 ++ ")"
+  terse (Addr _net cred (StakeRefPtr ptr)) = "Addr (" ++ terse cred ++ ") (" ++ terse ptr ++ ")"
+  terse (Addr _net cred StakeRefNull) = "Addr (" ++ terse cred ++ ") Null"
+  terse (AddrBootstrap x) = "BootStrap " ++ show x
+
+instance Terse (Credential keyrole c) where
+  terse (ScriptHashObj (ScriptHash hash)) = "Script " ++ show hash
+  terse (KeyHashObj (KeyHash hash)) = "Key " ++ show hash
+
+instance Terse Ptr where
+  terse (Ptr (SlotNo n) i j) = "Ptr " ++ show n ++ " " ++ show i ++ " " ++ show j
+
+instance Terse (TxId era) where
+  terse (TxId safehash) = show (extractHash safehash)
+
+instance CC.Crypto era => Terse (TxIn era) where
+  terse (TxIn txid n) = "In " ++ terse txid ++ " " ++ show n
+
+instance Terse (Coin) where
+  terse (Coin n) = show n
+
+{-
+tersediffincremental :: String -> IncrementalStake crypto -> IncrementalStake crypto -> String
+tersediffincremental message (IStake a b) (IStake c d) =
+   tersemapdiffs (message++" "++"hashes") a c ++
+   tersemapdiffs (message++" "++"ptrs") b d
+
+terseutxo :: (Era era,Terse (Core.TxOut era)) => String -> UTxO era -> String
+terseutxo message (UTxO mp) = terselist message (Map.toList mp)
+
+-}
```

Comment:
These are very usefulldebugging functions, which can't be defined here since UTxO and IncrementalStake, cannot be imported into this function. Someone will need them again, so I leave then here. so the source is not lost.

Reply:
In the end, I moved the whole file to the test-suite, so these can actually be defined.